src/pkg/oci/copier.go

```diff
@@ -6,50 +6,44 @@ package oci
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"time"
 
-	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/sync/semaphore"
+	"oras.land/oras-go/v2"
 )
 
-// CopyPackage copies a package from one OCI registry to another
-func CopyPackage(ctx context.Context, src *OrasRemote, dst *OrasRemote, include func(d ocispec.Descriptor) bool, concurrency int) error {
+// Copy copies an artifact from one OCI registry to another
+func Copy(ctx context.Context, src *OrasRemote, dst *OrasRemote,
+	include func(d ocispec.Descriptor) bool, concurrency int, progressBar helpers.ProgressWriter) (err error) {
+	if progressBar == nil {
+		progressBar = helpers.DiscardProgressWriter{}
+	}
 	// create a new semaphore to limit concurrency
 	sem := semaphore.NewWeighted(int64(concurrency))
 
 	// fetch the source root manifest
-	srcRoot, err := src.FetchRoot()
+	srcRoot, err := src.FetchRoot(ctx)
 	if err != nil {
 		return err
 	}
 
-	var layers []ocispec.Descriptor
-	for _, layer := range srcRoot.Layers {
-		if include != nil && include(layer) {
-			layers = append(layers, layer)
-		} else if include == nil {
-			layers = append(layers, layer)
-		}
-	}
-
+	layers := helpers.Filter(srcRoot.Layers, include)
 	layers = append(layers, srcRoot.Config)
 
-	size := int64(0)
-	for _, layer := range layers {
-		size += layer.Size
-	}
-
-	title := fmt.Sprintf("[0/%d] layers copied", len(layers))
-	progressBar := message.NewProgressBar(size, title)
-	defer progressBar.Successf("Copied %s", src.repo.Reference)
 	start := time.Now()
 
 	for idx, layer := range layers {
-		message.Debug("Copying layer:", message.JSONValue(layer))
+		b, err := json.MarshalIndent(layer, "", "  ")
+		if err != nil {
+			src.log.Debug("ERROR marshalling json: %s", err.Error())
+		}
+		src.log.Debug("Copying layer:", string(b))
 		if err := sem.Acquire(ctx, 1); err != nil {
```

Comment:
This seems like a lot of logging - it might be better to log the layer digest (the actual contents can be looked up later if needed)

Reply:
i liked the brrr when I was writing the first iteration of this func, + logging the entire descriptor tells you also size and mediatype, really useful for debugging. Plus this is logged _before_ the layer push which allows you to see if your error is on a really large layer.

Reply:
Yeah I don't mind it especially since it's debug